### PR TITLE
[LINQ] Replace MapOperator with FlatMapOperator, and add flat_map to API.

### PR DESCRIPTION
**Pull request by pschafhalter**
- Replaces `MapOperator` with a more general `FlatMapOperator`.
- Updates documentation.
- Fixes examples and tests.

### CODE REVIEW

File: erdos/src/dataflow/operators/map.rs [hunk 1]
```diff
@@ -10,77 +10,110 @@ use crate::dataflow::{
...
-            .unwrap();
-        tracing::debug!(
-            "{} @ {:?}: received {:?} and sent {:?}",
-            ctx.get_operator_config().get_name(),
-            ctx.get_timestamp(),
-            data,
-            (self.map_function)(data)
-        );
+    fn on_data(&mut self, ctx: &mut OneInOneOutContext<(), I::Item>, data: &D) {
+        tracing::info!("on data");
```
Comment (sukritkalra) [inline erdos/src/dataflow/operators/map.rs:64]: Any particular reason this was kept separately and on `info!(...)`?

File: erdos/src/dataflow/operators/map.rs [hunk 2]
```diff
@@ -10,77 +10,110 @@ use crate::dataflow::{
...
+            let timestamp = ctx.get_timestamp().clone();
+            let msg = Message::new_message(timestamp, item);
+            ctx.get_write_stream().send(msg).unwrap();
+        }
     }
 
-    fn on_watermark(&mut self, _ctx: &mut OneInOneOutContext<(), D2>) {}
+    fn on_watermark(&mut self, _ctx: &mut OneInOneOutContext<(), I::Item>) {
+        tracing::info!("{}: got watermark", _ctx.get_operator_config().get_name());
+    }
```
Comment (sukritkalra) [inline erdos/src/dataflow/operators/map.rs:82]:
```suggestion
    fn on_watermark(&mut self, ctx: &mut OneInOneOutContext<(), I::Item>) {
        tracing::info!("{}: got watermark", ctx.get_operator_config().get_name());
    }
```

File: erdos/src/dataflow/operators/map.rs [hunk 3]
```diff
@@ -95,7 +128,25 @@ where
...
+            self,
+        )
+    }
+
+    fn flat_map<F, I>(&self, flat_map_fn: F) -> Stream<D2>
+    where
+        F: 'static + Fn(&D1) -> I + Send + Sync + Clone,
+        I: 'static + IntoIterator<Item = D2>,
+    {
+        let op_name = format!("FlatMapOp_{}", self.name());
```
Comment (sukritkalra) [inline erdos/src/dataflow/operators/map.rs:146]:
```suggestion
        let op_name = format!("FlatMapOp_{}", self.id());
```

**Review by sukritkalra — CHANGES_REQUESTED** [review]

File: erdos/src/dataflow/operators/map.rs (hunk 1 above)
Comment (pschafhalter) [reply erdos/src/dataflow/operators/map.rs:64]: Sorry, those are debug statements I forgot to remove.

File: erdos/src/dataflow/operators/map.rs (hunk 3 above)
Comment (pschafhalter) [reply erdos/src/dataflow/operators/map.rs:146]: Why do we prefer to use the stream ID over the stream name?

Comment (sukritkalra) [reply erdos/src/dataflow/operators/map.rs:143]: Same problem as before, if you cascade operators via LINQ the name ends up being `MapOp_SplitOp_FilterOp_writestream_writestream_writestream`.

**Review by sukritkalra — APPROVED** [review]
LGTM, with the `id()` change.

**Pull request merged**